File: DSA/src/Questions/Graph/FrogPositionAfterTSeconds.java
package Questions.Graph;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
/*
https://leetcode.com/problems/frog-position-after-t-seconds/description/
 Given an undirected tree consisting of n vertices numbered from 1 to n. A frog starts jumping from vertex 1. In one second, the frog jumps from its current vertex to another unvisited vertex if they are directly connected. The frog can not jump back to a visited vertex. In case the frog can jump to several vertices, it jumps randomly to one of them with the same probability. Otherwise, when the frog can not jump to any unvisited vertex, it jumps forever on the same vertex.

The edges of the undirected tree are given in the array edges, where edges[i] = [ai, bi] means that exists an edge connecting the vertices ai and bi.

Return the probability that after t seconds the frog is on the vertex target. Answers within 10-5 of the actual answer will be accepted.

 

Example 1:


Input: n = 7, edges = [[1,2],[1,3],[1,7],[2,4],[2,6],[3,5]], t = 2, target = 4
Output: 0.16666666666666666 
Explanation: The figure above shows the given graph. The frog starts at vertex 1, jumping with 1/3 probability to the vertex 2 after second 1 and then jumping with 1/2 probability to vertex 4 after second 2. Thus the probability for the frog is on the vertex 4 after 2 seconds is 1/3 * 1/2 = 1/6 = 0.16666666666666666. 
Example 2:


Input: n = 7, edges = [[1,2],[1,3],[1,7],[2,4],[2,6],[3,5]], t = 1, target = 7
Output: 0.3333333333333333
Explanation: The figure above shows the given graph. The frog starts at vertex 1, jumping with 1/3 = 0.3333333333333333 probability to the vertex 7 after second 1. 
 

Constraints:

1 <= n <= 100
edges.length == n - 1
edges[i].length == 2
1 <= ai, bi <= n
1 <= t <= 50
1 <= target <= n
Time complexity: O(n)
Space complexity: O(n)
 */
public class FrogPositionAfterTSeconds {
    List<Integer>[] g;

    public double frogPosition(int n, int[][] edges, int t, int target) {
        if (n == 1) {
            return 1.0;
        }
        g = new ArrayList[n + 1];
        boolean[] visited = new boolean[n + 1];
        Arrays.setAll(g, k -> new ArrayList<>());
        for (int[] e : edges) {
            g[e[0]].add(e[1]);
            g[e[1]].add(e[0]);
        }
        return helper(1, t, target, visited);
    }

    public double helper(int node, int t, int target, boolean[] visited) {
        if (node != 1 && g[node].size() == 1 || t == 0) {
            if (node == target) {
                return 1.0;
            }
            return 0.0;
        }
        visited[node] = true;
        double res = 0.0;
        for (int next : g[node]) {
            if (visited[next]) {
                continue;
            }
            res += helper(next, t - 1, target, visited);
            if (res > 0) {
                if (node != 1) {
                    return res / (g[node].size() - 1);
                } else {
                    return res / g[node].size();
                }
            }
        }
        return 0.0;
    }
}

File: DSA/Python/Questions/Graph/frog_position_after_t_seconds.py
"""
# Question: Frog Position After T Seconds
# Link: https://leetcode.com/problems/frog-position-after-t-seconds/

# Problem Statement:
# Given an undirected tree consisting of n vertices numbered from 1 to n. A frog starts jumping from vertex 1. In one second, the frog jumps from its current vertex to another unvisited vertex if they are directly connected. The frog can only jump to an unvisited vertex. The frog cannot jump back to a visited vertex. In case the frog cannot jump to any unvisited vertex, it will stay on the current vertex.
# Return the probability that after t seconds the frog is on the vertex target.

# Example:
# Input: n = 7, edges = [[1,2],[1,3],[1,7],[2,4],[2,6],[3,5]], t = 2, target = 4
# Output: 0.16666666666666666
"""

from typing import List
from collections import defaultdict

# DFS Solution
"""
Algorithm:
1. Build an adjacency list from the edges.
2. Use DFS to explore the tree starting from vertex 1.
3. Track the probability of reaching each vertex.
4. Return the probability of reaching the target vertex after t seconds.

Time Complexity: O(V + E) where V is the number of vertices and E is the number of edges.
Space Complexity: O(V) for the visited array and recursion stack.
"""


def frog_position(n: int, edges: List[List[int]], t: int, target: int) -> float:
    def dfs(node: int, time: int, prob: float) -> float:
        # Base case: if time is up or we reach the target
        if time == 0 or node == target:
            return prob if node == target else 0

        visited[node] = True  # Mark the current node as visited
        # Find all unvisited neighbors
        unvisited_neighbors = [
            neighbor for neighbor in graph[node] if not visited[neighbor]
        ]
        num_unvisited = len(unvisited_neighbors)

        # If there are no unvisited neighbors, check if we're at the target
        if num_unvisited == 0:
            return prob if node == target else 0

        # Explore each unvisited neighbor
        for neighbor in unvisited_neighbors:
            result = dfs(neighbor, time - 1, prob / num_unvisited)
            if result > 0:
                return result

        visited[node] = False  # Unmark the node if not successful
        return 0

    # Build the graph as an adjacency list
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)
        graph[v].append(u)

    visited = [False] * (n + 1)  # Track visited nodes
    return dfs(1, t, 1.0)  # Start DFS from node 1 with full probability


def main():
    test_cases = [
        {
            "n": 7,
            "edges": [[1, 2], [1, 3], [1, 7], [2, 4], [2, 6], [3, 5]],
            "t": 2,
            "target": 4,
            "expected": 0.16666666666666666,
        },
        {
            "n": 7,
            "edges": [[1, 2], [1, 3], [1, 7], [2, 4], [2, 6], [3, 5]],
            "t": 1,
            "target": 7,
            "expected": 0.3333333333333333,
        },
        {"n": 3, "edges": [[2, 1], [3, 2]], "t": 1, "target": 2, "expected": 1.0},
    ]

    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(
            f"n: {test['n']}, edges: {test['edges']}, t: {test['t']}, target: {test['target']}"
        )
        result = frog_position(test["n"], test["edges"], test["t"], test["target"])
        print(f"Result: {result}, Expected: {test['expected']}")
        assert abs(result - test["expected"]) < 1e-9, f"Test case {i + 1} failed"


if __name__ == "__main__":
    main()
